### PR TITLE
refactor: abstract tenant-specific settings 

### PR DESCRIPTION
**Pull request by kangmingtay**
## What kind of change does this PR introduce?
* This PR attempts to decouple tenant-specific configuration before starting the API server

### CODE REVIEW

File: internal/api/tenant.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+		}).SetBurst(30),
+		// limit per 5 minutes
+		"sso": tollbooth.NewLimiter(config.RateLimitSso/(60*5), &limiter.ExpirableOptions{
+			DefaultExpirationTTL: time.Hour,
+		}).SetBurst(30),
+		// limit per 5 minutes
+		"sso/assertion": tollbooth.NewLimiter(config.SAML.RateLimitAssertion/(60*5), &limiter.ExpirableOptions{
+			DefaultExpirationTTL: time.Hour,
+		}).SetBurst(30),
+	}
```
Comment (hf) [inline internal/api/tenant.go:68]: Do these have to be a keyed map? Why not a struct with named fields?

File: internal/api/middleware.go [hunk 2]
```diff
@@ -46,9 +44,10 @@ func (f *FunctionHooks) UnmarshalJSON(b []byte) error {
 	return nil
 }
 
-func (a *API) limitHandler(lmt *limiter.Limiter) middlewareHandler {
+func (a *API) limitHandler(rateLimiterType string) middlewareHandler {
 	return func(w http.ResponseWriter, req *http.Request) (context.Context, error) {
 		c := req.Context()
+		lmt := a.limiters[rateLimiterType]
```
Comment (hf) [inline internal/api/middleware.go:50]: Let's use named struct fields or maybe a named enum type instead?

File: internal/api/middleware_test.go [hunk 3]
```diff
@@ -185,7 +181,7 @@ func (ts *MiddlewareTestSuite) TestLimitEmailOrPhoneSentHandler() {
 			req.Header.Set("Content-Type", "application/json")
 			w := httptest.NewRecorder()
 
-			for i := 0; i < 5; i++ {
+			for i := 0; i < 30; i++ {
```
Comment (J0) [inline internal/api/middleware_test.go:184]: /nit would be nice to specify where this number comes from (e.g. bursts) or define a constant for it

File: internal/api/rate_limiters.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+		// limit per minute
+		MFAVerify: tollbooth.NewLimiter(config.MFA.RateLimitChallengeAndVerify/60, &limiter.ExpirableOptions{
+			DefaultExpirationTTL: time.Minute,
+		}).SetBurst(30),
+		// limit per minute
+		MFAChallenge: tollbooth.NewLimiter(config.MFA.RateLimitChallengeAndVerify/60, &limiter.ExpirableOptions{
+			DefaultExpirationTTL: time.Minute,
+		}).SetBurst(30),
+		// limit per 5 minutes
+		SSO: tollbooth.NewLimiter(config.RateLimitSso/(60*5), &limiter.ExpirableOptions{
```
Comment (J0) [inline internal/api/rate_limiters.go:60]: We should probably add the SSO rate limits into the [production readiness guide](https://supabase.com/docs/guides/platform/going-into-prod#rate-limits) once we release SSO for Apps

File: internal/api/tenant.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	config   *conf.GlobalConfiguration
+	limiters *Limiters
+}
+
+func (t *Tenant) GetConnection() *storage.Connection {
+	return t.db
+}
+
+func NewTenant(config *conf.GlobalConfiguration) (*Tenant, error) {
+	// TODO: close db connection in cleanup
```
Comment (J0) [inline internal/api/tenant.go:23]: qq: what needs to be happen before we can close these TODOs (e.g. further refactoring, infra changes)?

**Comment by kangmingtay** [comment]
closing for now since this needs further discussion

**Pull request closed without merging**